File: java/dtdgen-master/src/main/java/com/philhanna/dtdgen/DocumentModel.java
package com.philhanna.dtdgen;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeMap;

/**
 * Provides access to what is known about the structure of a set of XML
 * document instances
 */
public class DocumentModel {

   // ====================================================================
   // Class constants and variables
   // ====================================================================

   // ====================================================================
   // Class methods
   // ====================================================================

   // ====================================================================
   // Instance variables
   // ====================================================================
   /**
    * Returns the root element name. This is determined by looking at
    * all elements in the element name list and deleting of their known
    * child elements. What remains should be the root element name (if
    * the XML is well-formed)
    */
   public String getRootElementName() {

      // Make a list of the names of all element that are not known to
      // have parent elements. To begin with, this will be every element

      final Set<String> possibleRoots = new HashSet<String>();
      for (final Iterator<String> it = elementNameIterator(); it.hasNext();) {
         final String elementName = it.next();
         possibleRoots.add(elementName);
      }

      // Now systematically remove names from this list by looking
      // through all elements and eliminating their children.

      for (final Iterator<String> it = elementNameIterator(); it.hasNext();) {

         // Get an element name

         final String parentElementName = it.next();
         final ElementModel parentElementModel = getElementModel(
               parentElementName);

         // How many child elements does it have?

         final int nChildren = parentElementModel.getChildModelCount();

         // Get each child element and remove its name from the
         // possibleRoots list

         for (int index = 0; index < nChildren; index++) {
            final ChildModel childModel = parentElementModel
                  .getChildModel(index);
            final String childElementName = childModel.getName();
            possibleRoots.remove(childElementName);
         }
      }

      // There should be only one element left - get the count

      final List<String> rootCandidates = new ArrayList<String>();
      rootCandidates.addAll(possibleRoots);
      final int nRootCandidates = rootCandidates.size();

      // Too few

      if (nRootCandidates == 0) {
         throw new RuntimeException("No possible root elements found");
      }

      // Too many

      if (nRootCandidates > 1) {
         final StringBuilder sb = new StringBuilder();
         for (int i = 0; i < nRootCandidates; i++) {
            if (i > 0)
               sb.append(", ");
            sb.append(rootCandidates.get(i));
         }
         final String rootCandidateNames = sb.toString();

         final String errmsg = String.format(
               "%d possible root elements found: [%s]",
               nRootCandidates,
               rootCandidateNames);
         throw new RuntimeException(errmsg);
      }

      // Found the one possible root element

      final String rootElementName = rootCandidates.get(0);
      return rootElementName;
   }

   /**
    * Alphabetical list of element types appearing in the document; each
    * has the element name as a key and an ElementModelImpl object as
    * the value.
    */
   private final Map<String, ElementModel> elementMap = new TreeMap<String, ElementModel>();

   // ====================================================================
   // Constructors
   // ====================================================================

   // ====================================================================
   // Implementation of DocumentModel
   // ====================================================================

   /**
    * Returns the {@link Foo} with the given name
    * @param elementName the element name
    * @return the element model with that name, or <code>null</code>, if
    *         it does not exist in the document model
    */
   public ElementModel getElementModel(String elementName) {
      return elementMap.get(elementName);
   }

   /**
    * Returns an iterator over the list of element model names
    */
   public Iterator<String> elementNameIterator() {
      return elementMap.keySet().iterator();
   }

   // ====================================================================
   // Instance methods
   // ====================================================================

   public void addElementModel(ElementModel elementModel) {
      final String elementName = elementModel.getName();
      elementMap.put(elementName, elementModel);
   }

}

File: dtdgen/document_model.py
from dtdgen import ElementModel


class DocumentModel:
    """Provides access to what is known about the structure of a set of XML
    document instances"""

    def __init__(self):
        self._element_map: dict[str, ElementModel] = dict()

    def get_root_element_name(self) -> str:
        """Returns the root element name. This is determined by looking at
        all elements in the element name list and deleting all their known
        child elements. What remains should be the root element name (if
        the XML is well-formed)"""

        # Make a set of the names of all elements that are not known to have
        # parent elements.  To begin with, this will be every element
        possible_roots: set[str] = {name for name in self._element_map}

        # Now systematically remove names from this set by looking through
        # all elements and eliminating their children from the set
        for parent_element_name, parent_element_model in self._element_map.items():
            for child in parent_element_model.child_iterator():
                child_element_name = child.name
                possible_roots.remove(child_element_name)

        # There should be only one element left - get the count
        # and raise an exception if there are too few or too many
        root_candidates: list[str] = list(possible_roots)
        if not root_candidates:
            raise RuntimeError("No possible root elements found")
        if len(root_candidates) > 1:
            errmsg = f"{len(root_candidates)} possible root elements found: "
            errmsg += ", ".join(root_candidates)
            raise RuntimeError(errmsg)

        root_element_name = root_candidates[0]
        return root_element_name

    def add_element_model(self, element_model: ElementModel):
        """Adds an element model to this document model"""
        element_name = element_model.name
        self._element_map[element_name] = element_model

    def get_element_model(self, name: str) -> ElementModel:
        """Returns the element model with the specificd name,
        or None, if it does not exist in the document model"""
        return self._element_map.get(name)

    def element_name_iterator(self):
        """Returns an iterator over the list of element model names"""
        for name in self._element_map:
            yield name
